mintscript/src/script/mod.rs

```diff
@@ -0,0 +1,232 @@
+
+    /// Construct a threshold.
+    ///
+    /// Like [Self::new] but panics instead of failing if an invalid threshold is created.
+    const fn new_unchecked(required: usize, conditions: Vec<ScriptCondition>) -> Self {
+        Self {
+            required,
+            conditions,
+        }
+    }
+
+    /// Construct a threshold.
+    ///
+    /// Creates an N-of-K threshold construct where N out of K conditions have to be satisfied.
+    /// Fails if N > K.
+    pub fn new(
+        required: usize,
+        conditions: Vec<ScriptCondition>,
+    ) -> Result<Self, ScriptConstructionError> {
+        ensure!(
+            required <= conditions.len(),
+            ScriptConstructionError::InvalidThreshold(required, conditions.len()),
+        );
+        Ok(Self::new_unchecked(required, conditions))
+    }
+
+    /// A [WitnessScript] consisting of only this threshold.
+    pub const fn into_script(self) -> WitnessScript {
+        WitnessScriptInner::Threshold(self).into_script()
+    }
+
+    /// Get all the conditions in this threshold construct.
+    pub fn conditions(&self) -> &[ScriptCondition] {
+        &self.conditions
+    }
+
+    /// Number of conditions that need to be satisfied in order for the threshold to be valid.
+    pub fn required(&self) -> usize {
+        self.required
+    }
+
+    /// Number of conditions that need to be dissatisfied in order for the threshold to be valid.
+    pub fn required_dissat(&self) -> usize {
+        self.conditions().len() - self.required()
+    }
+
+    /// Collect conditions that the prover claims to satisfy.
+    pub fn collect_satisfied_unchecked(&self) -> Vec<&WitnessScript> {
+        self.conditions.iter().filter_map(ScriptCondition::as_satisfied).collect()
+    }
+
+    /// Collect conditions that the prover claims to satisfy and check the number of these meets
+    /// the threshold requirements.
+    pub fn collect_satisfied(&self) -> Result<Vec<&WitnessScript>, ThresholdError> {
+        // Track how many conditions are there to be left satisfied and dissatisfied
+        let mut left_sat = self.required();
+        let mut left_dissat = self.required_dissat();
+        let mut satisfied = Vec::with_capacity(left_sat);
+
+        for cond in self.conditions() {
+            match cond {
+                ScriptCondition::Satisfied(ws) => {
+                    left_sat = left_sat.checked_sub(1).ok_or(ThresholdError::Excessive)?;
+                    satisfied.push(ws);
+                }
+                ScriptCondition::Dissatisfied(_) => {
+                    left_dissat = left_dissat.checked_sub(1).ok_or(ThresholdError::Insufficient)?;
+                }
+            }
+        }
+
+        assert_eq!(left_dissat, 0);
+        assert_eq!(left_sat, 0);
+
+        Ok(satisfied)
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+enum WitnessScriptInner {
+    Threshold(Threshold),
+    Signature(Destination, InputWitness),
+    Timelock(OutputTimeLock),
+}
```

Comment:
I wonder whether we should add htlc now so that we ensure no design issues will appear